packages/react-router/lib/components.tsx

```diff
@@ -214,18 +218,24 @@ export function Navigate({
       `only ever rendered in response to some user interaction or state change.`
   );
 
-  let dataRouterState = React.useContext(DataRouterStateContext);
+  let { matches } = React.useContext(RouteContext);
+  let { pathname: locationPathname } = useLocation();
   let navigate = useNavigate();
 
-  React.useEffect(() => {
-    // Avoid kicking off multiple navigations if we're in the middle of a
-    // data-router navigation, since components get re-rendered when we enter
-    // a submitting/loading state
-    if (dataRouterState && dataRouterState.navigation.state !== "idle") {
-      return;
-    }
-    navigate(to, { replace, state, relative });
-  });
+  // Resolve the path outside of the effect so that when effects run twice in
+  // StrictMode they navigate to the same place
+  let path = resolveTo(
+    to,
+    getPathContributingMatches(matches).map((match) => match.pathnameBase),
+    locationPathname,
+    relative === "path"
+  );
+  let jsonPath = JSON.stringify(path);
+
+  React.useEffect(
+    () => navigate(JSON.parse(jsonPath), { replace, state, relative }),
+    [navigate, jsonPath, relative, replace, state]
+  );
```

Comment:
The reason this was problematic in 6.11 is that we would call navigate twice and when in a `RouterProvider` we'd delay resolving the relative path until inside the router.  And thus the second execution would re-resolve `to` against the current location (which in a on-loader scenario would already be updated from the first effect).

Instead, we resolve the absolute path here in `Navigate` so that duplicate calls to navigate via the data router go to the same path - just as they do in `BrowserRouter`.

Reply:
I'm a little worried about `state` being in there. I'd still like to see this be an empty array as I think we are going to see bad app/library provider code that immediately flows new state / props down causing a new state identity due to no-one ever in the history of app code memoing the value they pass in. If tests are passing across react v17 and 18 though I'm find shipping it.

Reply:
For future reference we tested the unit test setups from [93ccb2b](https://github.com/remix-run/react-router/pull/10435/commits/93ccb2b5d967142270a084cdd166bcc34943fefb) in demo apps using react 16.8 and 18 and confirmed the UI was the same (even if the underlying React.StrictMode execution approaches differed).  We may try to see if we can get our test suite running against both versions in a separate undertaking.

Reply:
Also, I think we're no worse off than we were previously where the effect had no second param since it would have re-run every time anyway?

Reply:
Thanks!

Reply:
> I'm a little worried about `state` being in there. I'd still like to see this be an empty array as I think we are going to see bad app/library provider code that immediately flows new state / props down causing a new state identity due to no-one ever in the history of app code memoing the value they pass in. If tests are passing across react v17 and 18 though I'm fine shipping it.

This wound up biting me. I had an object that I was passing to state that wasn't memo'd. I think the check for dataRouterState !== "idle" saved me prior to 6.11.1, but with this change I had a flurry of redirects (each calling a loader and reloading a file hundreds of times).